README.md
# PX-CHART IS NOW DEPRECATED
## It has been replaced by px-vis-timeseries: [https://github.com/PredixDev/px-vis-timeseries](https://github.com/PredixDev/px-vis-timeseries)

# px-chart

Element that defines a chart, using sub-elements for axis, series and controls configuration.

## Overview

### px-chart

#### Minimal Usage - 1 series of data

    <px-chart>
      <px-chart-series-line id="hpt-acc-position-cruise" data="{{...}}">
      </px-chart-series-line>
    </px-chart>

```js
// Expected data format
[
  [1397102460000, 11.4403],
  [1397139660000, 13.1913],
  [1397177400000, 12.8485],
  [1397228040000, 10.975],
  [1397248260000, 12.9377],
  [1397291280000, 13.3795],
  [1397318100000, 13.0869],
  [1397342100000, 17.3758]
]
// or
[
  {'x': 1397102460000, 'y': 11.4403},
  {'x': 1397139660000, 'y': 13.1913},
  {'x': 1397177400000, 'y': 12.8485},
  {'x': 1397228040000, 'y': 10.975},
  {'x': 1397248260000, 'y': 12.9377},
  {'x': 1397291280000, 'y': 13.3795},
  {'x': 1397318100000, 'y': 13.0869},
  {'x': 1397342100000, 'y': 17.3758}
]
```

#### Full configuration - multiple series, axis, chart controls

	<px-chart tooltip-type="condensed" series-events-width="1">
	    <px-chart-controls data-controls show-date-range="true"></px-chart-controls>
	    <px-chart-yaxis title='{"text": "Left Axis"}' id="firstAxis" offset="0"></px-chart-yaxis>
	    <px-chart-yaxis title='{"text": "Right Axis"}' id="anotherAxis" opposite="true" offset="-2.5"></px-chart-yaxis>
	    <px-chart-series-line id="delta-egt-cruise" axis-id="firstAxis" upper-threshold="30.5" lower-threshold="6.25" data="{{myData}}">
	    </px-chart-series-line>
	    <px-chart-series-line id="fan-vibration-cruise" axis-id="anotherAxis" data="{{myData2}}">
	    </px-chart-series-line>
	  </px-chart>

#### Different types of series

    <px-chart-series-line>
    <px-chart-series-scatter>
    <px-chart-series-histogram>
    <px-chart-series-bar>

#### Dynamically adding series

      var myChart = document.querySelector("#my-chart");

      fetchMyData().then(function(firstSeries) {
        firstSeries.id = firstSeries.name; // you must have an id (you can just set it equal to the series name)
        myChart.addSeries(firstSeries);
      });

```js
// Sample response
{
  "name": "delta-egt-cruise",
  "series": [
    [1397102460000, 11.4403],
    [1397139660000, 13.1913],
    [1397177400000, 12.8485],
    [1397228040000, 10.975],
    [1397248260000, 12.9377],
    [1397291280000, 13.3795],
    [1397318100000, 13.0869],
    [1397342100000, 17.3758]
  ]
}
```

#### Attributes

##### type

*Type:* **String** - (*Optional*) - *Default:* "line"

Determines that what type of chart to display.  This type influences the series type for the navigator, the x-axis display, etc.

This does NOT affect the type of the series displayed on the graph.  To change the type of series, change the ```<px-chart-series-*>``` web component to line, bar, histogram, etc.

Can only be statically configured (not data-bindable).

Accepts values of 'line', 'scatter' or 'bar'.

```
<px-chart
	...
	type="bar">
</px-chart>
```

##### navigatorDisabled

*Type:* **Boolean** - (*Optional*) - *Default:* false

Whether to show the zoom-able / scroll-able area at the bottom of the chart.

Can only be statically configured (not data-bindable).

```
<px-chart
	...
	navigator-disabled>
</px-chart>
```

##### chartState

*Type:* **Object** - (*Optional*) - *Default:* {}

Holds chart state for binding between charts & serialising chart settings.

See the synchronized charts demo (demo-sync.html) for an example.

```
<px-chart
	...
	chart-state="{{chartState}}">
</px-chart>
```

##### dataVisColors

*Type:* **Object** - (*Optional*) - *Default:* same as datavis colors in px-colors-design

Mapping of color name to rgb value for use in datavis (axis, navigator, series, etc. colors)

Can only be statically configured (not data-bindable).

```
<px-chart
	...
	data-vis-colors="{...}">
</px-chart>
```


##### exportServerUrl

*Type:* **String** - (*Optional*) - *Default:* null

URL for the chart export server (converts to image / pdf / etc). Default is null. Can use "http://export.highcharts.com" for demo purposes only...no intellectual property should go through that server.

Can only be statically configured (not data-bindable).

```
<px-chart
	...
	export-server-url="https://my-export-url.ge.com">
</px-chart>
```


##### height

*Type:* **Number** - (*Optional*) - *Default:* 400

See [http://api.highcharts.com/highcharts#chart.height](http://api.highcharts.com/highcharts#chart.height)

Can only be statically configured (not data-bindable).

```
<px-chart
	...
	height="500">
</px-chart>
```


##### legend

*Type:* **Object** - (*Optional*) - *Default:* (has default)

See [http://api.highcharts.com/highcharts#legend](http://api.highcharts.com/highcharts#legend)

Note that a default legend will be enabled but can set this as an override.

Can only be statically configured (not data-bindable).
```
<px-chart
	...
	legend='{
     "enabled": true,
     "useHTML": true,
     "layout": "vertical",
     "verticalAlign": "middle",
     "y": 44,
     "align": "left",
     "floating": true,
     "itemMarginTop": 5,
     "itemMarginBottom": 5
     }'>
</px-chart>
```


##### legendRight

*Type:* **Boolean** - (*Optional*) - *Default:* false

Whether the legend appears to the right of the chart.

Can only be statically configured (not data-bindable).

```
<px-chart
	...
	legend-right>
</px-chart>
```

##### margin

*Type:* **Array** - (*Optional*) - *Default:* [50,50,30,50] or [100,50,30,50] if zoom buttons exist

See [http://api.highcharts.com/highcharts#chart.margin](http://api.highcharts.com/highcharts#chart.margin)

Can only be statically configured (not data-bindable).
```
<px-chart
	...
	margin="[...]">
</px-chart>
```

##### plotBorderWidth

*Type:* **Number** - (*Optional*) - *Default:* 1

See [http://api.highcharts.com/highcharts#chart.plotBorderWidth](http://api.highcharts.com/highcharts#chart.plotBorderWidth)

Can only be statically configured (not data-bindable).
```
<px-chart
	...
	plot-border-width="3">
</px-chart>
```

##### seriesColorOrder

*Type:* **Array** - (*Optional*) - *Default:* [dv-basic-blue, dv-basic-orange, dv-basic-green, dv-basic-pink, dv-basic-brown, dv-basic-purple, dv-basic-yellow, dv-basic-red, dv-basic-gray, dv-light-blue, dv-light-orange, dv-light-green, dv-light-pink, dv-light-brown, dv-light-purple, dv-light-yellow, dv-light-red, dv-light-gray, dv-dark-blue, dv-dark-orange, dv-dark-green, dv-dark-pink, dv-dark-brown, dv-dark-purple, dv-dark-yellow, dv-dark-red, dv-dark-gray]

Mapping of color names (from dataVisColors map) in the order they should be applied to chart series.

Can only be statically configured (not data-bindable).

```
<px-chart
	...
	series-color-order='[
      "dv-light-pink",
      "dv-light-brown",
      "dv-light-purple",
      "dv-light-yellow"
    ]'>
</px-chart>
```

##### seriesEventsColor

*Type:* **String** - (*Optional*) - *Default:* "rgb(59,59,63)"

The color of the vertical line drawn for series events. Any valid CSS color string is supported.

Can only be statically configured (not data-bindable).

```
<px-chart
	...
	series-events-color="rgb(33,33,22)">
</px-chart>
```
##### seriesEventsWidth

*Type:* **Number** - (*Optional*) - *Default:* 2

The width of the vertical line drawn for series events.

Can only be statically configured (not data-bindable).
```
<px-chart
	...
	series-events-width="5">
</px-chart>
```
##### spacingBottom

*Type:* **Number** - (*Optional*) - *Default:* 20

See [http://api.highcharts.com/highstock#chart.spacingBottom](http://api.highcharts.com/highstock#chart.spacingBottom)

Can only be statically configured (not data-bindable).
```
<px-chart
	...
	spacing-bottom="13">
</px-chart>
```
##### tooltipKind

*Type:* **String** - (*Optional*) - *Default:* "px"

Selects the kind of tooltip to display
* px: default type with tooltip appearing in a row above the chart
* hc: tooltip follows cursor

See the charts demo (demo.html) for an example.

Can only be statically configured (not data-bindable).
```
<px-chart
	...
	tooltip-type="condensed">
</px-chart>
```
##### tooltipType

*Type:* **String** - (*Optional*) - *Default:* "normal"

Selects the chart tooltip type

See the charts demo (demo.html) for an example. Can select between normal and condensed.

Can only be statically configured (not data-bindable).
```
<px-chart
	...
	tooltip-type="condensed">
</px-chart>
```
##### tooltipOffset

*Type:* **Number** - (*Optional*) - *Default:* 0

Sets the pixel distance of the tooltip from the top of the chart plot area.

If no value is supplied, a 40px for condensed and 60px for normal tooltips offset is applied.
```
<px-chart
	...
	tooltip-offset="20">
</px-chart>
```
##### tooltipTimestamp

*Type:* **String** - (*Optional*) - *Default:* 'show'

Valid values are: 'show' & 'hide'

Shows or hides the timestamp in the tooltip
```
<px-chart
	...
	tooltip-timestamp="hide">
</px-chart>
```
##### tooltipTimeFormat

*Type:* **String** - (*Optional*) - *Default:* 'HH:mm:ss ZZ'

Configures how the time is displayed in the normal tooltip

See http://momentjs.com/docs/#/parsing/string-format/ for String formats
```
<px-chart
	...
	tooltip-time-format="HH:mm:ss ZZ">
</px-chart>
```
##### tooltipDateFormat

*Type:* **String** - (*Optional*) - *Default:* 'MMM DD, YYYY'

Configures how the date is displayed in the normal tooltip

See http://momentjs.com/docs/#/parsing/string-format/ for String formats
```
<px-chart
	...
	tooltip-date-format="DD MMM YYYY">
</px-chart>
```
##### tooltipDatetimeFormat

*Type:* **String** - (*Optional*) - *Default:* 'HH:mm:ss ZZ | DD MMM YYYY'

Configures how the time and date are displayed in the condensed tooltip

See http://momentjs.com/docs/#/parsing/string-format/ for String formats
```
<px-chart
	...
	tooltip-datetime-format="HH:mm:ss - MM/DD/YY">
</px-chart>
```
##### zoomControls

*Type:* **String** - (*Optional*) - *Default:* "showcontrols"

Valid values are: 'hidecontrols' & 'showcontrols'

Specifies if zoom should open Predix custom overlay with controls or just run default zoom.
Running default zoom allows for more zoomTypes as specified below.

```
<px-chart
	...
	zoom-controls="hidecontrols">
</px-chart>
```
##### zoomType

*Type:* **String** - (*Optional*) - *Default:* "x"

See [http://api.highcharts.com/highcharts#chart.zoomType](http://api.highcharts.com/highcharts#chart.zoomType)

Can only be statically configured (not data-bindable).

```
<px-chart
	...
	zoom-type="xy">
</px-chart>
```

<br />
<hr />

#### Methods
##### addSeries(seriesConfig, noRedraw)

seriesConfig Object
* {String} id
* {Array} data
* {Number} yAxis Optional. The axis index to which the series should be bound. Defaults to 0.
* {Number} lineWidth Optional.
* {Object} marker. Optional. Highcharts marker config
* {Object} tooltip. Optional. Highcharts tooltip config
* {Boolean} noRedraw. Optional. If true, does not force a chart redraw() after adding or updating the series

Adds a series to the chart, adding a yAxis as needed

##### addYAxis(axisConfig, defaultColor, noRedraw)

Add a y-axis with the given config object.

##### hasSeries(seriesId) : Boolean

Returns true if the chart has a series with the given id

##### refreshAllSeries()

Fires one "refresh-series" event for each series on the chart, each event has the id of the series.

##### removeSeries(seriesId)

Removes a series from the chart

##### togglePointMarkers()
Can pass optional list of seriesIds, or if null than toggles all point markers.

Toggles display of points on the chart

##### updateAxisThreshold(seriesConfig, thresholdValue, id)

Threshold lines are optionally bound to series. This function processes threshold values on a given series and applies them on the y-axis that is associated with the given series.

##### updateSeries(seriesId, data, noRedraw)

Updates a series on the chart, adding a default series as needed and redrawing after series is added unless noRedraw specified..

<br />
<hr />

### px-chart-series-*

Element defining a chart series that is displayed as a line.

Can be: px-chart-series-line, px-chart-series-bar, px-chart-series-scatter, or px-chart-series-histogram.

#### Usage

##### Data

    <px-chart-series-line data="{{...}}" id="hpt-acc-position-cruise">
    </px-chart-series-line>

    <px-chart-series-line name="The Display Name" line-width="5" lower-threshold="12" upper-threshold="16" id="hpt-acc-position-cruise" marker='{"enabled": false, "radius": 15}' data="{{...}}">
    </px-chart-series-line>

```js
// Expected data format:
[
  [1397102460000, 11.4403],
  [1397139660000, 13.1913],
  [1397177400000, 12.8485],
  [1397228040000, 10.975]
]
// or
[
  {'x': 1397102460000, 'y': 11.4403},
  {'x': 1397139660000, 'y': 13.1913},
  {'x': 1397177400000, 'y': 12.8485},
  {'x': 1397228040000, 'y': 10.975},
  {'x': 1397248260000, 'y': 12.9377},
  {'x': 1397291280000, 'y': 13.3795},
  {'x': 1397318100000, 'y': 13.0869},
  {'x': 1397342100000, 'y': 17.3758}
]
```

##### Series Object

    <px-chart-series-line id="my-series"
      series-obj-name-key="myName"
      series-obj-data-key="myData">
    </px-chart-series-line>

```js
  document.querySelector('#my-series').seriesObj = {
          myData: [
              [1397102460000, 11.4403],
              [1397139660000, 13.1913],
              [1397177400000, 12.8485],
              [1397228040000, 10.975]],
          myName: 'foo'
      };
  ```

##### Using iron-ajax

    <px-chart-series-line id="..." upper-threshold="0.75" lower-threshold="0.25">
      <iron-ajax
          url="../demo-data/aviation/fan-vibration-cruise.json"
          handleas="json"></iron-ajax>
    </px-chart-series-line>

<br />
<hr />

#### Attributes

##### id

*Type:* **String** - (*Required:*)

Unique id of this series

Can only be statically configured (not data-bindable).

```
<px-chart-series-line
  ...
  id="my-series">
</px-chart-series-line>
```

##### axisId

*Type:* **String** - (*Optional*)

Optional declaration of which y axis id this series should be plotted against, if more than one.

Can only be statically configured (not data-bindable).

```
<px-chart-series-line
	...
	axis-id="first-axis">
</px-chart-series-line>
```

##### data

*Type:* **Array** - (*Optional*)

The data associated with the series, an Array of Arrays with the first value of each inner Array being a unix timestamp and the second value being the value.

Fires a "data-changed" event via the "notify" flag when the data changes, the assumption is that the chart will listen for this event.

```
<px-chart-series-line
	...
	data="{{myData}}">
</px-chart-series-line>
```

##### dataEvents

*Type:* **Array** - (*Optional*)

Events associated with the series, an Array of Objects with the keys:

* "time" {Number}, a unix timestamp of the event time
* "id" {String}, unique id of the event
* "label" {String}, the text label associated with the event

Fires a "dataEvents-changed" event via the "notify" flag when the events changes, the assumption is that the chart will listen for this event.

```
<px-chart-series-line
	...
	data-events="[...]">
</px-chart-series-line>
```

##### lineWidth

*Type:* **Number** - (*Optional*) - *Default:* 1

The width of the line for this series

Can only be statically configured (not data-bindable).

```
<px-chart-series-line
	...
	line-width="3">
</px-chart-series-line>
```

##### lowerThreshold

*Type:* **Number** - (*Optional*)

Lower value beyond which the data should be shown as "above threshold".

Can only be statically configured (not data-bindable).

```
<px-chart-series-line
	...
	lower-threshold="24">
</px-chart-series-line>
```

##### marker

*Type:* **Object** - (*Optional*) - *Default:* {"enabled": false, "radius": 2}

Config for each plotted point on the line

Can only be statically configured (not data-bindable).

```
<px-chart-series-line
	...
	marker='{"enabled": false, "radius": 15}'>
</px-chart-series-line>
```

##### name

*Type:* **String** - (*Optional*) - *Default:* id

Display name of the series, defaults to id

Can only be statically configured (not data-bindable).

```
<px-chart-series-line
	...
	name="my-series-name">
</px-chart-series-line>
```

##### units

*Type:* **String** - (*Optional*) - *Default:* ''

Set the units of the series. It can then be displayed in the chart tooltip

Can only be statically configured (not data-bindable).

```
<px-chart-series-line
	...
	units="morgens">
</px-chart-series-line>
```

##### seriesObj

*Type:* **Object** - (*Optional*)

Optional object that contains members that map to #name and #data via the seriesObjDataKey and seriesObjNameKey,

e.g. seriesObj[seriesObjDataKey] and seriesObj[seriesObjNameKey]

```
<px-chart-series-line
	...
	series-obj="{{...}}">
</px-chart-series-line>
```

##### seriesObjDataEventsKey

*Type:* **String** - (*Optional*) - *Default:* "events"

Key in the optional seriesObj to be used for #events

Can only be statically configured (not data-bindable)

```
<px-chart-series-line
	...
	series-obj-data-events-key="my-events">
</px-chart-series-line>
```
##### seriesObjDataKey

*Type:* **String** - (*Optional*) - *Default:* "series"

Key in the optional seriesObj to be used for #data

Can only be statically configured (not data-bindable).

```
<px-chart-series-line
	...
	series-obj-data-key="my-data">
</px-chart-series-line>
```
##### seriesObjNameKey

*Type:* **String** - (*Optional*) - *Default:* "name"

Key in the optional seriesObj to be used for #name

Can only be statically configured (not data-bindable).

```
<px-chart-series-line
	...
	series-obj-name-key="my-series-name">
</px-chart-series-line>
```
##### tooltip

*Type:* **Object** - (*Optional*) - *Default:* {"valueDecimals": 2}

Config for the tooltip for each point

Can only be statically configured (not data-bindable).

```
<px-chart-series-line
	...
	tooltip='{"valueDecimals": 3}'>
</px-chart-series-line>
```
##### upperThreshold

*Type:* **Number** - (*Optional*)

Upper value beyond which the data should be shown as "above threshold".

Can only be statically configured (not data-bindable).

```
<px-chart-series-line
	...
	upper-threshold="250">
</px-chart-series-line>
```

<br />
<hr />

### px-chart-yaxis

Element to define y-axis configuration.

Maps to Highcharts yaxis config, see docs at [http://api.highcharts.com/highcharts#yAxis](http://api.highcharts.com/highcharts#yAxis)

#### Usage

    <px-chart-yaxis attr-1="..." attr-2="..."></px-chart-yaxis>

<br />
<hr />

#### Attributes

##### id

*Type:* **String** - (*Required*)

Unique id of this axis, primarily for associating series to it

Can only be statically configured (not data-bindable).

```
<px-chart-yaxis
  ...
  id="firstAxis">
</px-chart-yaxis>
```

##### dateTimeLabelFormats

*Type:* **Object** - (*Optional*)

[http://api.highcharts.com/highcharts#yAxis.dateTimeLabelFormats](http://api.highcharts.com/highcharts#yAxis.dateTimeLabelFormats)

Can only be statically configured (not data-bindable).

```
<px-chart-yaxis
	...
	date-time-label-formats="{...}">
</px-chart-yaxis>
```

##### defaultColorProps

*Type:* **Array** - (*Optional*) - *Default:* ["tickColor", "lineColor"]

List of property names in the axis config that should be given a default color (based on the index of this axis as passed from the containing chart).

Can only be statically configured (not data-bindable).

```
<px-chart-yaxis
	...
	default-color-props="[...]">
</px-chart-yaxis>
```

##### gridLineWidth

*Type:* **Number** - (*Optional*)

[http://api.highcharts.com/highcharts#yAxis.gridLineWidth](http://api.highcharts.com/highcharts#yAxis.gridLineWidth)

Can only be statically configured (not data-bindable).

```
<px-chart-yaxis
	...
	grid-line-width="37">
</px-chart-yaxis>
```

##### labels

*Type:* **Object** - (*Optional*) - *Default:* {"style": {"fontSize": 0.8rem, "color": #000000}, "y": 5}


HISTORY.md
v1.8.5
==================
* Deprecated

v1.8.4
==================
* Updated deprecation warning

v1.8.3
==================
* Updating so px-demo-snippet and px-api-viewer get new grays

v1.8.2
==================
* Update colors design to pick up new colors

v1.8.1
==================
* changing ghp.sh to account for Alpha releases

v1.8.0
==================
* Updated dependencies

v1.7.8
==================
* changing browser in wct testing from safari 8 to safari 10 on elcapitan

v1.7.7
==================
* changing all devDeps to ^

v1.7.6
==================
* Update px-theme to 2.0.1 and update test fixtures

v1.7.5
==================
* update dependencies for dropdown

v1.7.4
==================
* removing px-theme style call


v1.7.3
==================
* changing Gruntfile.js to gulpfile.js


v1.7.2
==================
* bower updating px-demo-snippet

v1.7.1
=================
* Update dependencies

v1.6.7
=================
* fixed bower conflicts

v1.6.6
=================
* fixed inception bug in index file with vix-timeseries link

v1.6.5
=================
* added deprecation warning

v1.6.4
=================
* added mega demo, fixed spelling mistake in ghp.sh

v1.6.3
=================
* Updated vulcanize index and demo

v1.6.2
=================
* Added vulcanize

v1.6.1
=================
* Load moment from px-datetime-imports

v1.6.0
=================
* Upgrade to Polymer 1.5.0

v1.5.3
=================
* added oss_notice to bower ignore

v1.5.2
=================
* added pull request test for travis and updated OSS Notice

v1.5.1
=================
* added auto-github pages functionality

v1.5.0
=================
* Upgrade to Polymer 1.4.0

v1.4.1
=================
* updated url for demo json

v1.4.0
=================
* Added tooltipKind attribute which allows the dev to decide on the kind of tooltip to display: px tooltip in a row above the chart or hc tooltip which follows the cursor

=================
v1.3.0
=================
* Added time and date stamp to tooltip with developer set configuration attributes on px-chart
* Added units attribute to all px-chart-series components which gets shown in the tooltip
* refactored tooltip code

v1.2.1
=================
* Updated License

v1.2.0
==================
* Added attr zoom-controls to px-chart which allows the dev to specify if they want to use the custom, Predix zoom with overlay and controls, or the default Highcharts zoom. Default then allows zoomType on x,y,xy.

v0.0.1
==================
* Initial release


px-time-series.js
Polymer({

    is: 'px-time-series',

    initialSeries: null,

    /**
     * Properties block, expose attribute values to the DOM via 'reflect'
     *
     * @property properties
     * @type Object
     */
    properties: {
        queries: {
            type: Array,
            notify: true,
            observer: 'queriesChanged'
        },
        rangeStart: {
            type: String,
            observer: 'rangeStartUpdated'
        },
        rangeEnd: {
            type: String,
            observer: 'rangeEndUpdated'
        },
        showyaxisunits: {
            type: Boolean
        },
        xAxisLabel: {
            type: String
        },
        yAxisLabel: {
            type: String
        }
    },

    ready: function() {
        var chartConfig = this.buildConfig();
        this.chart = new Highcharts.StockChart(chartConfig);
        if(this.initialSeries) {
            this.updateChartSeries(this.initialSeries);
        }
    },
    rangeStartUpdated: function () {
        this.rangeStartStr = this.getDateStr(this.rangeStart);
    },
    rangeEndUpdated: function () {
        this.rangeEndStr = this.getDateStr(this.rangeEnd);
    },

    buildConfig: function() {
        var self = this;

        this.getRenderEl = this.getRenderEl || function() {
            return this.$.container;
        };

        var monochromePalette = {
            'black': '#141414',
            'grayDarkest': '#2b2b2b',
            'grayDarker': '#414141',
            'grayDark': '#575757',
            'gray': '#868686',
            'grayLight': '#bcbcbc',
            'grayLighter': '#d4d4d4',
            'grayLightest': '#e9e9e9',
            'offwhite': '#f5f5f5',
            'white': '#fff'
        };

        var accentPalette = {
            'cyan': '#005CB9',
            'orange': '#ff9821',
            'green': '#46ad00',
            'purple': '#8669ff',
            'red': '#de2533',
            'yellow': '#ffed45',
            'cyanLight': '#3693f8',
            'orangeLight': '#ffbb66',
            'greenLight': '#75d835',
            'purpleLight': '#9c97ff',
            'redLight': '#ff5c5c',
            'yellowDark': '#ffcf45',
            'cyanDark': '#00366e',
            'orangeDark': '#e55c00',
            'greenDark': '#1d5f11',
            'purpleDark': '#595194',
            'redDark': '#b61225',
            'yellowLight': '#fff98d',
            'blue': '#005bb8',
            'blueLight': '#3693f8',
            'blueDark': '#00366e'
        };

        var PXd = {
            'accentPalette': accentPalette,
            'monochromePalette': monochromePalette
        };

        var convertMapToValueArray = function(map){
            var valArray = [];
            for(var key in map) {
                if(map.hasOwnProperty(key)) {
                    valArray.push(map[key]);
                }
            }
            return valArray;
        };

        var config = {
            annotationsOptions: {
              enabledButtons: false
            },
            chart: {
                backgroundColor: "transparent",
                events: {
                    redraw: function() {
                        var extremes = this.xAxis[0].getExtremes();
                        self.rangeStart = extremes.min;
                        self.rangeEnd = extremes.max;
                    }
                },
                height: 400,
                margin: [90,30,30,30],
                plotBorderWidth: 2,
                renderTo: this.getRenderEl(),
                spacing: [0,0,25,0],
                style: {
                    fontFamily: 'inherit',
                    fontSize: 'inherit'
                },
                zoomType: 'x'
            },
            colors: convertMapToValueArray(PXd.accentPalette),
            credits: {
                enabled: false
            },
            legend: {
                align: 'left',
                enabled: true,
                floating: true,
                itemMarginBottom: 10,
                itemStyle: {
                  "fontSize": "inherit",
                  "fontWeight": "normal"
                },
                margin: 0,
                padding: 0,
                symbolPadding: 5,
                symbolWidth: 10,
                verticalAlign: 'top'
            },
            navigation: {
                buttonOptions: {
                  enabled: false
                }
            },
            navigator: {
                adaptToUpdatedData: false,
                height: 50,
                margin: 15,
                outlineColor: PXd.monochromePalette.gray,
                maskInside: true,
                series: {
                    color: 'transparent',
                    lineColor: PXd.accentPalette.blue,
                    lineWidth: 2
                },
                xAxis: {
                  labels: {
                    style: {
                      fontSize: "0.8rem"
                    },
                    y: 15
                  }
                },
                xAxis: {
                  gridLineWidth: 0
                }
            },
            plotOptions: {
                line: {
                    lineWidth: 1,
                    states: {
                        hover: {
                            lineWidth: 1
                        }
                    }
                },
                scatter: {
                    marker: {
                        enabled: true
                    }
                },
                series: {
                    marker: {}
                }
            },
            rangeSelector: {
                enabled: false
            },
            series: [],
            scrollbar: {
                enabled: false
            },
            title: {
                text: null
            },
            tooltip: {
                backgroundColor: PXd.monochromePalette.white,
                borderColor: PXd.monochromePalette.grayLighter,
                shadow: false,
                style: {
                    fontFamily: 'inherit',
                    fontSize: 'inherit'
                },
                headerFormat: '<span>{point.key}</span><br/>',
                pointFormat: '<span style="color:{series.color}">{series.name}: {point.y}</span><br/>'
            },
            xAxis: {
                events: {
                    afterSetExtremes: function(event) {
                        self.fire('after-set-extremes', event);
                    }
                },
                labels: {
                  align: "left",
                  style: {
                    fontSize: '0.8rem'
                  },
                  x: 3,
                  y: 12
                },
                startOnTick: true,
                title: {
                    text: null
                }
            },
            yAxis: {
                labels: {
                  style: {
                    fontSize: '0.8rem'
                  },
                  y: 5
                },
                opposite: false,
                startOnTick: true,
                tickWidth: 2,
                title: {
                    text: null
                }
            }
        };

        return config;
    },
    queriesChanged: function(newData, oldData) {
        if (!newData) {
            return;
        }

        var seriesToShow = this.dataTransform(newData);

        // if the chart is not initialized yet... don't do any of this, just store series for later
        if(this.chart === undefined) {
            this.initialSeries = seriesToShow;
            return;
        }

        this.updateChartSeries(seriesToShow);
    },
    updateChartSeries: function(seriesToShow) {
        var newIds = _.pluck(seriesToShow, 'name');
        var currentIds = _.pluck(this.chart.series, 'name');
        newIds.push('Navigator'); // HACK: Need 'Navigator' series to exist in the newIds so we do not remove it.

        // Get ids of series that we will be touching
        var idsToUpdate = _.intersection(currentIds, newIds);
        var idsToRemove = _.difference(currentIds, newIds);
        var idsToAdd = _.difference(newIds, currentIds);

        var self = this;

        // Update series that already exist
        _.each(idsToUpdate, function(idToUpdate) {
            _.each(seriesToShow, function(series) {
                if (series.name === idToUpdate) {
                    self.chart.get(idToUpdate).setData(series.data);
                }
            });
        });

        // Remove old ones
        _.each(idsToRemove, function(idToRemove) {
            self.chart.get(idToRemove).remove();
        });

        // Add new series
        _.each(idsToAdd, function(idToAdd) {
            _.each(seriesToShow, function(series) {
                if (series.name === idToAdd) {
                    self.addSeries(idToAdd, series.data);
                }
            });
        });

        this.chart.reflow();
    },
    addSeries: function(seriesId, data) {

        var newseries = {
            id: seriesId,
            name: seriesId,
            data: data
        };

        this.chart.addSeries(newseries);
    },
    dataTransform: function(newData) {
        var highchartSeries = [];
        var self = this;
        if (newData && newData.constructor === Array) {
            //series
            newData.forEach(function(query) {
                if (query.results && query.results.constructor === Array) {
                    query.results.forEach(function(result) {
                        //validate result format
                        if (result.name && result.values && result.values.constructor === Array) {
                            highchartSeries.push({
                                name: result.name,
                                data: result.values
                            });
                        }
                        else {
                            console.warn('Series data is missing name or values property');
                        }
                    });
                }
                else {
                    //TODO : add error and angular.emit at the base Directive level
                    console.error('Invalid time series data format');
                }
            });
        }
        else {
            console.error('Invalid time series data format');
        }

        return highchartSeries;
    },
    hasExtremeChanged: function (rangeStart, rangeEnd) {
        var extremes = this.chart.xAxis[0].getExtremes();
        return extremes.min !== rangeStart || extremes.max !== rangeEnd;
    },
    getDateStr: function (d) {
        function ensureTwoDigits(s) {
            s = '' + s;
            while (s.length < 2) {
                s = '0' + s;
            }
            return s;
        }

        var date = new Date(d);

        if (!date || !date.getHours || isNaN(date.getTime())) {
            return '';
        }
        return (ensureTwoDigits(date.getHours())) + ':' + ensureTwoDigits(date.getMinutes()) +
            ' ' + (date.getMonth() + 1) + '/' + (date.getDate()) + '/' + (date.getFullYear());
    },
    setMonthsOfRange: function (event, detail, sender) {
        var months = parseInt(event.target.getAttribute('data-month'));

        this.rangeEnd = this.rangeEnd || Date.now();

        var temp = new Date(this.rangeEnd);
        var month = temp.getMonth() - months;
        temp.setMonth(month);

        this.rangeStart = temp.getTime();

        this.setExtremesIfChanged(this.rangeStart, this.rangeEnd);
    },
    setRangeToYTD: function () {
        this.rangeEnd = Date.now();

        var tempStartDate = new Date();
        tempStartDate.setMonth(0);
        tempStartDate.setDate(1);
        tempStartDate.setHours(0);
        tempStartDate.setMinutes(0);
        tempStartDate.setSeconds(0);

        this.rangeStart = tempStartDate.getTime();
        this.setExtremesIfChanged(this.rangeStart, this.rangeEnd);
    },
    submitHandler: function () {
        if (this.isValidDate(this.rangeStartStr, true) && this.isValidDate(this.rangeEndStr, true)) {
            var startTime = new Date(this.rangeStartStr).getTime();
            var endTime = new Date(this.rangeEndStr).getTime();
            this.setExtremesIfChanged(startTime, endTime);
        }
    },
    setExtremesIfChanged: function (startTime, endTime) {
        if (this.hasExtremeChanged(startTime, endTime)) {
            this.rangeStart = startTime;
            this.rangeEnd = endTime;
            this.chart.xAxis[0].setExtremes(this.rangeStart, this.rangeEnd);
        }

        // always set the visible strings back to a good value
        this.rangeStartStr = this.getDateStr(this.rangeStart);
        this.rangeEndStr = this.getDateStr(this.rangeEnd);
    },
    getRangeClasses: function (s) {
        var isValid = this.isValidDate(s);
        return isValid ? 'text-input' : 'invalid-date text-input';
    },
    isValidDate: function (s, checkForNull) {
        if (!checkForNull && !s) {
            return true;
        }
        var re = /^[012]?\d:[012345]\d\s+(0?[1-9]|1[012])\/(0?[1-9]|[1-2]\d|3[01])\/[12][90]\d\d$/;
        return re.test(s);
    }//,
//            vDestroy: function () {
//                if (this.chart) {
//                    this.chart.destroy();
//                }
////                this.vElement.remove();
//            }

});
